File: backend/internal/dto/racao.go
package dto

import (
	"time"

	"github.com/google/uuid"
)

// CreateFeedReceiptRequest representa o payload de registro de recebimento de racao.
type CreateFeedReceiptRequest struct {
	TipoRacaoID     *uuid.UUID `json:"tipo_racao_id,omitempty"`
	QuantidadeKg    float64    `json:"quantidade_kg" validate:"required,gt=0"`
	DataRecebimento string     `json:"data_recebimento" validate:"required"` // formato: YYYY-MM-DD
	Fornecedor      *string    `json:"fornecedor,omitempty" validate:"omitempty,max=200"`
	LoteRacao       *string    `json:"lote_racao,omitempty" validate:"omitempty,max=100"`
	Origem          string     `json:"origem,omitempty" validate:"omitempty,oneof=compra remanescente_anterior sobra_final"`
}

// FeedReceiptResponse representa a resposta de um recebimento de racao.
type FeedReceiptResponse struct {
	ID              uuid.UUID  `json:"id"`
	LoteID          uuid.UUID  `json:"lote_id"`
	TipoRacaoID     *uuid.UUID `json:"tipo_racao_id,omitempty"`
	TipoRacaoNome   string     `json:"tipo_racao_nome,omitempty"`
	QuantidadeKg    float64    `json:"quantidade_kg"`
	DataRecebimento time.Time  `json:"data_recebimento"`
	Fornecedor      *string    `json:"fornecedor,omitempty"`
	LoteRacao       *string    `json:"lote_racao,omitempty"`
	Origem          string     `json:"origem"`
	CreatedAt       time.Time  `json:"created_at"`
}

// CreateFeedConsumptionRequest representa o payload de registro de consumo de racao.
type CreateFeedConsumptionRequest struct {
	Data         string  `json:"data" validate:"required"` // formato: YYYY-MM-DD
	QuantidadeKg float64 `json:"quantidade_kg" validate:"required,gt=0"`
}

// FeedConsumptionResponse representa a resposta de um consumo de racao.
type FeedConsumptionResponse struct {
	ID           uuid.UUID `json:"id"`
	LoteID       uuid.UUID `json:"lote_id"`
	Data         time.Time `json:"data"`
	QuantidadeKg float64   `json:"quantidade_kg"`
	CreatedAt    time.Time `json:"created_at"`
}

// SaldoRacaoResponse representa o saldo atual de racao de um lote.
type SaldoRacaoResponse struct {
	RecebidoTotalKg    float64  `json:"recebido_total_kg"`
	ConsumoTotalKg     float64  `json:"consumo_total_kg"`
	SaldoKg            float64  `json:"saldo_kg"`
	DiasEstoqueRestante *float64 `json:"dias_estoque_restante,omitempty"`
}

File: flutter_app/lib/features/lote_detail/domain/entities/racao.dart
/// Registro de recebimento de racao para um lote.
///
/// Campos mapeados a partir do endpoint GET /lotes/{id}/feed_receipts
/// e do DTO FeedReceiptResponse do backend Go.
class RecebimentoRacao {
  final String id;
  final String loteId;
  final String dataRecebimento;

  /// Quantidade recebida em kg.
  final double quantidadeKg;

  /// ID do tipo de racao (opcional).
  final String? tipoRacaoId;

  /// Nome do tipo de racao (opcional, retornado pelo backend via join).
  final String? tipoRacaoNome;

  /// Fornecedor da racao (opcional).
  final String? fornecedor;

  /// Numero do lote da racao (opcional).
  final String? loteRacao;

  /// Origem do recebimento: 'compra', 'remanescente_anterior', 'sobra_final'.
  final String origem;

  final String createdAt;

  const RecebimentoRacao({
    required this.id,
    required this.loteId,
    required this.dataRecebimento,
    required this.quantidadeKg,
    this.tipoRacaoId,
    this.tipoRacaoNome,
    this.fornecedor,
    this.loteRacao,
    required this.origem,
    required this.createdAt,
  });

  factory RecebimentoRacao.fromJson(Map<String, dynamic> json) {
    return RecebimentoRacao(
      id: json['id'] as String,
      loteId: json['lote_id'] as String,
      dataRecebimento: json['data_recebimento'] as String,
      quantidadeKg: (json['quantidade_kg'] as num).toDouble(),
      tipoRacaoId: json['tipo_racao_id'] as String?,
      tipoRacaoNome: json['tipo_racao_nome'] as String?,
      fornecedor: json['fornecedor'] as String?,
      loteRacao: json['lote_racao'] as String?,
      origem: json['origem'] as String? ?? 'compra',
      createdAt: json['created_at'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'lote_id': loteId,
      'data_recebimento': dataRecebimento,
      'quantidade_kg': quantidadeKg,
      if (tipoRacaoId != null) 'tipo_racao_id': tipoRacaoId,
      if (tipoRacaoNome != null) 'tipo_racao_nome': tipoRacaoNome,
      if (fornecedor != null) 'fornecedor': fornecedor,
      if (loteRacao != null) 'lote_racao': loteRacao,
      'origem': origem,
      'created_at': createdAt,
    };
  }
}

/// Registro de consumo diario de racao de um lote.
///
/// Campos mapeados a partir do endpoint GET /lotes/{id}/feed_consumptions
/// e do DTO FeedConsumptionResponse do backend Go.
class ConsumoRacao {
  final String id;
  final String loteId;
  final String data;

  /// Quantidade consumida em kg.
  final double quantidadeKg;

  final String createdAt;

  const ConsumoRacao({
    required this.id,
    required this.loteId,
    required this.data,
    required this.quantidadeKg,
    required this.createdAt,
  });

  factory ConsumoRacao.fromJson(Map<String, dynamic> json) {
    return ConsumoRacao(
      id: json['id'] as String,
      loteId: json['lote_id'] as String,
      data: json['data'] as String,
      quantidadeKg: (json['quantidade_kg'] as num).toDouble(),
      createdAt: json['created_at'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'lote_id': loteId,
      'data': data,
      'quantidade_kg': quantidadeKg,
      'created_at': createdAt,
    };
  }
}
